File: frontend/src/types/projects.ts
/**
 * Types for multi-project office state.
 * Matches backend models in app/models/projects.py.
 */

import type { Agent, Boss, OfficeState, TodoItem } from "./generated";

export type ViewMode =
  | "office"
  | "projects"
  | "project"
  | "sessions"
  | "session";

export interface ProjectGroup {
  key: string;
  name: string;
  color: string;
  root: string | null;
  agents: Agent[];
  boss: Boss;
  sessionCount: number;
  todos: TodoItem[];
}

export interface MultiProjectGameState {
  sessionId: string;
  projects: ProjectGroup[];
  office: OfficeState;
  lastUpdated: string;
}

/**
 * Short display name: last directory segment from root path.
 * Falls back to stripping "Projects-" prefix from name when root is missing.
 */
export function getProjectDisplayName(project: { name: string; root: string | null }): string {
  if (project.root) {
    const segments = project.root.replace(/\/+$/, "").split("/");
    const last = segments[segments.length - 1];
    if (last) return last;
  }
  const name = project.name;
  if (name.startsWith("Projects-")) return name.slice("Projects-".length);
  return name;
}

export interface ProjectSummary {
  key: string;
  name: string;
  color: string;
  root: string | null;
  sessionCount: number;
}

File: backend/app/models/projects.py
"""Models for multi-project office state."""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field
from pydantic.alias_generators import to_camel

from app.models.agents import Agent, Boss, OfficeState
from app.models.common import TodoItem


class ProjectGroup(BaseModel):
    """A project room containing agents from all sessions of that project."""

    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)

    key: str
    name: str
    color: str
    root: str | None
    agents: list[Agent]
    boss: Boss
    session_count: int
    todos: list[TodoItem] = Field(default_factory=list)


class MultiProjectGameState(BaseModel):
    """Complete state grouped by project for the multi-room view."""

    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)

    session_id: str = "__all__"
    projects: list[ProjectGroup]
    office: OfficeState
    last_updated: datetime
